### PR TITLE
[Improve][Connector-V2][Kafka] Support to specify multiple partition keys

### PR DESCRIPTION
**Pull request by harveyyue**
## Check list

* [ ] Code changed are covered with tests, or it does not need tests for reason:
* [ ] If any new Jar binary package adding in your PR, please add License Notice according
  [New License Guide](https://github.com/apache/incubator-seatunnel/blob/dev/docs/en/contribution/new-license.md)
* [ ] If necessary, please update the documentation to describe the new feature. https://github.com/apache/incubator-seatunnel/tree/dev/docs

### CODE REVIEW

File: docs/en/connector-v2/sink/Kafka.md [hunk 1]
```diff
@@ -51,7 +51,7 @@ In AT_LEAST_ONCE, producer will wait for all outstanding messages in the Kafka b
 
 NON does not provide any guarantees: messages may be lost in case of issues on the Kafka broker and messages may be duplicated.
 
-### partition_key [string]
+### partition_key [array]
 
 Configure which field is used as the key of the kafka message.
```
Comment (EricJoy2048) [inline docs/en/connector-v2/sink/Kafka.md:57]: Please add `changed logs` reference https://github.com/apache/incubator-seatunnel/blob/dev/docs/en/connector-v2/source/HdfsFile.md

File: seatunnel-connectors-v2/connector-kafka/src/main/java/org/apache/seatunnel/connectors/seatunnel/kafka/sink/KafkaSinkWriter.java [hunk 2]
```diff
@@ -162,11 +170,24 @@ private Properties getKafkaProperties(Config pluginConfig) {
...
+            for (int i = 0; i < seaTunnelRowType.getFieldNames().length; i++) {
+                if (this.partitionKeys.contains(seaTunnelRowType.getFieldNames()[i])) {
+                    keyFieldNames[index] = seaTunnelRowType.getFieldName(i);
+                    keyFieldTypes[index] = seaTunnelRowType.getFieldType(i);
+                    ++index;
+                }
+            }
+            SeaTunnelRowType keySeaTunnelRowType = new SeaTunnelRowType(keyFieldNames, keyFieldTypes);
+            return new DefaultSeaTunnelRowSerializer(this.topic, keySeaTunnelRowType, seaTunnelRowType);
+        } else {
```
Comment (EricJoy2048) [inline seatunnel-connectors-v2/connector-kafka/src/main/java/org/apache/seatunnel/connectors/seatunnel/kafka/sink/KafkaSinkWriter.java:150]: Suggest 
```
if () {
   return xx;
} else if () {
  return xx;
} else {
  return xx;
}
```

replace to

```
if () {
    return xx;
}

if () {
  return xx;
}

return xx;
```

File: seatunnel-connectors-v2/connector-kafka/src/main/java/org/apache/seatunnel/connectors/seatunnel/kafka/sink/KafkaSinkWriter.java [hunk 3]
```diff
@@ -188,23 +209,39 @@ private void restoreState(List<KafkaSinkState> states) {
...
         };
     }
+
+    private List<String> createPartitionKeys(Config pluginConfig, SeaTunnelRowType seaTunnelRowType) {
+        if (pluginConfig.hasPath(PARTITION_KEY)) {
+            return pluginConfig.getStringList(PARTITION_KEY).stream()
+                    .filter(f -> {
+                        if (Arrays.asList(seaTunnelRowType.getFieldNames()).contains(f)) {
+                            return true;
+                        } else {
```
Comment (EricJoy2048) [inline seatunnel-connectors-v2/connector-kafka/src/main/java/org/apache/seatunnel/connectors/seatunnel/kafka/sink/KafkaSinkWriter.java]: `else` is redundant.

**Comment by harveyyue** [comment]
@EricJoy2048 Done.

**Review by EricJoy2048 — DISMISSED** [review]
LGTM @hailin0  PTAL

File: docs/en/connector-v2/sink/Kafka.md [hunk 4]
```diff
@@ -113,3 +113,4 @@ sink {
 ### 2.3.0-beta 2022-10-20
 
 - Add Kafka Sink Connector
+- Support to specify multiple partition keys
```
Comment (EricJoy2048) [inline docs/en/connector-v2/sink/Kafka.md]:
```suggestion
### next version

- Support to specify multiple partition keys
```

File: docs/en/connector-v2/sink/Kafka.md [hunk 5]
```diff
@@ -113,3 +113,6 @@ sink {
 ### 2.3.0-beta 2022-10-20
 
 - Add Kafka Sink Connector
+### next version
+
+- Support to specify multiple partition keys
```
Comment (EricJoy2048) [inline docs/en/connector-v2/sink/Kafka.md]:
```suggestion
- [Feature] Support to specify multiple partition keys [3230](https://github.com/apache/incubator-seatunnel/pull/3230)
```

**Review by EricJoy2048 — DISMISSED** [review]
LGTM

File: seatunnel-e2e/seatunnel-connector-v2-e2e/connector-kafka-e2e/src/test/java/org/apache/seatunnel/e2e/connector/kafka/KafkaIT.java [hunk 6]
```diff
@@ -84,51 +83,55 @@ public void startKafkaContainer() {
...
+        }
+        if (kafkaContainer != null) {
+            kafkaContainer.close();
+        }
+    }
+
+    @TestTemplate
+    public void testSinkKafka(TestContainer container) throws IOException, InterruptedException {
+        Container.ExecResult execResult = container.executeJob("/kafkasink_fake_to_kafka.conf");
+        Assertions.assertEquals(0, execResult.getExitCode());
```
Comment (ic4y) [inline seatunnel-e2e/seatunnel-connector-v2-e2e/connector-kafka-e2e/src/test/java/org/apache/seatunnel/e2e/connector/kafka/KafkaIT.java]: Thank you for your contribution. In addition to checking that the job exits properly, you also need to check that data is successfully written to kafka

File: seatunnel-e2e/seatunnel-connector-v2-e2e/connector-kafka-e2e/src/test/java/org/apache/seatunnel/e2e/connector/kafka/KafkaIT.java [hunk 7]
```diff
@@ -84,51 +83,55 @@ public void startKafkaContainer() {
...
+
+    @TestTemplate
+    public void testSourceKafkaTextToConsole(TestContainer container) throws IOException, InterruptedException {
+        TextSerializationSchema serializer = TextSerializationSchema.builder()
+                .seaTunnelRowType(SEATUNNEL_ROW_TYPE)
+                .delimiter(",")
+                .build();
+        generateTestData(row -> new ProducerRecord<>("test_topic_text", null, serializer.serialize(row)));
+        Container.ExecResult execResult = container.executeJob("/kafkasource_text_to_console.conf");
+        Assertions.assertEquals(0, execResult.getExitCode(), execResult.getStderr());
```
Comment (ic4y) [inline seatunnel-e2e/seatunnel-connector-v2-e2e/connector-kafka-e2e/src/test/java/org/apache/seatunnel/e2e/connector/kafka/KafkaIT.java:159]: Ditto,  it is suggest to use AssertSink or fileSink to check whether the output data is as expected

File: seatunnel-e2e/seatunnel-connector-v2-e2e/connector-kafka-e2e/src/test/java/org/apache/seatunnel/e2e/connector/kafka/KafkaIT.java [hunk 8]
```diff
@@ -84,51 +83,55 @@ public void startKafkaContainer() {
...
+    }
+
+    @TestTemplate
+    public void testSourceKafkaJsonToConsole(TestContainer container) throws IOException, InterruptedException {
+        DefaultSeaTunnelRowSerializer serializer = new DefaultSeaTunnelRowSerializer("test_topic_json", SEATUNNEL_ROW_TYPE);
+        generateTestData(row -> serializer.serializeRow(row));
+        Container.ExecResult execResult = container.executeJob("/kafkasource_json_to_console.conf");
+        Assertions.assertEquals(0, execResult.getExitCode(), execResult.getStderr());
```
Comment (ic4y) [inline seatunnel-e2e/seatunnel-connector-v2-e2e/connector-kafka-e2e/src/test/java/org/apache/seatunnel/e2e/connector/kafka/KafkaIT.java:167]: Ditto

**Comment by hailin0** [comment]
@TaoZex please help to review

File: seatunnel-e2e/seatunnel-connector-v2-e2e/connector-kafka-e2e/src/test/resources/kafkasink_fake_to_kafka.conf [hunk 9]
```diff
@@ -0,0 +1,68 @@
...
+}
+
+transform {
+}
+
+sink {
+  Kafka {
+    bootstrap.servers = "kafkaCluster:9092"
+    topic = "test_topic"
+    partition_key = ["c_map","c_string"]
```
Comment (hailin0) [inline seatunnel-e2e/seatunnel-connector-v2-e2e/connector-kafka-e2e/src/test/resources/kafkasink_fake_to_kafka.conf]: How to use a fixed string partition key value?  eg: partition_key = 'aaa'

Comment (hailin0) [reply seatunnel-e2e/seatunnel-connector-v2-e2e/connector-kafka-e2e/src/test/resources/kafkasink_fake_to_kafka.conf]: It is recommended to implement in this way, which is simpler for users to use

partition_key = "${c_map},${c_string}"

Comment (hailin0) [reply seatunnel-e2e/seatunnel-connector-v2-e2e/connector-kafka-e2e/src/test/resources/kafkasink_fake_to_kafka.conf]: The value of partition_key can be a normal string or an expression to extract the field content. We have already supported single field extraction before

https://github.com/apache/incubator-seatunnel/pull/3085

Comment (harveyyue) [reply seatunnel-e2e/seatunnel-connector-v2-e2e/connector-kafka-e2e/src/test/resources/kafkasink_fake_to_kafka.conf]: @hailin0 You want to keep the logic of following code snippet, right?
`
        if (!fieldNames.contains(partitionKey)) {
            return row -> partitionKey;
        }
`

File: seatunnel-e2e/seatunnel-connector-v2-e2e/connector-kafka-e2e/src/test/java/org/apache/seatunnel/e2e/connector/kafka/KafkaIT.java (hunk 7 above)
Comment (harveyyue) [reply seatunnel-e2e/seatunnel-connector-v2-e2e/connector-kafka-e2e/src/test/java/org/apache/seatunnel/e2e/connector/kafka/KafkaIT.java:159]: Already use AssertSink

File: seatunnel-e2e/seatunnel-connector-v2-e2e/connector-kafka-e2e/src/test/java/org/apache/seatunnel/e2e/connector/kafka/KafkaIT.java (hunk 6 above)
Comment (harveyyue) [reply seatunnel-e2e/seatunnel-connector-v2-e2e/connector-kafka-e2e/src/test/java/org/apache/seatunnel/e2e/connector/kafka/KafkaIT.java]: Add the consumer data size and key message schema check

File: seatunnel-e2e/seatunnel-connector-v2-e2e/connector-kafka-e2e/src/test/java/org/apache/seatunnel/e2e/connector/kafka/KafkaIT.java (hunk 8 above)
Comment (harveyyue) [reply seatunnel-e2e/seatunnel-connector-v2-e2e/connector-kafka-e2e/src/test/java/org/apache/seatunnel/e2e/connector/kafka/KafkaIT.java:167]: Already use AssertSink

File: seatunnel-e2e/seatunnel-connector-v2-e2e/connector-kafka-e2e/src/test/resources/kafkasink_fake_to_kafka.conf (hunk 9 above)
Comment (hailin0) [reply seatunnel-e2e/seatunnel-connector-v2-e2e/connector-kafka-e2e/src/test/resources/kafkasink_fake_to_kafka.conf]: yes

Comment (harveyyue) [reply seatunnel-e2e/seatunnel-connector-v2-e2e/connector-kafka-e2e/src/test/resources/kafkasink_fake_to_kafka.conf]: BTW, the fixed string partition key value will cause to write all of data to a partition of topic, specify or omit partition key(s) is good practice.

Comment (hailin0) [reply seatunnel-e2e/seatunnel-connector-v2-e2e/connector-kafka-e2e/src/test/resources/kafkasink_fake_to_kafka.conf]: I agree

I'm worried about breaking the released feature

Comment (hailin0) [reply seatunnel-e2e/seatunnel-connector-v2-e2e/connector-kafka-e2e/src/test/resources/kafkasink_fake_to_kafka.conf]: Perhaps rename config key to `partition_key_fields` is more appropriate

Comment (hailin0) [reply seatunnel-e2e/seatunnel-connector-v2-e2e/connector-kafka-e2e/src/test/resources/kafkasink_fake_to_kafka.conf]: and need to update the examples & description in the docs

image

Comment (harveyyue) [reply seatunnel-e2e/seatunnel-connector-v2-e2e/connector-kafka-e2e/src/test/resources/kafkasink_fake_to_kafka.conf]: done.

File: seatunnel-e2e/seatunnel-connector-v2-e2e/connector-kafka-e2e/src/test/resources/log4j.properties [hunk 10]
```diff
@@ -0,0 +1,22 @@
+#
```
Comment (hailin0) [inline seatunnel-e2e/seatunnel-connector-v2-e2e/connector-kafka-e2e/src/test/resources/log4j.properties]: Remove this file.

This is the common configuration of e2e
https://github.com/apache/incubator-seatunnel/blob/dev/seatunnel-e2e/seatunnel-e2e-common/src/test/resources/log4j2.properties

File: seatunnel-e2e/seatunnel-connector-v2-e2e/connector-kafka-e2e/src/test/java/org/apache/seatunnel/e2e/connector/kafka/KafkaIT.java [hunk 11]
```diff
@@ -0,0 +1,248 @@
...
+
+    private KafkaContainer kafkaContainer;
+
+    @BeforeAll
+    @Override
+    public void startUp() throws Exception {
+        kafkaContainer = new KafkaContainer(DockerImageName.parse("confluentinc/cp-kafka:6.2.1"))
+                .withNetwork(NETWORK)
+                .withNetworkAliases(KAFKA_HOST)
+                .withLogConsumer(new Slf4jLogConsumer(log));
```
Comment (hailin0) [inline seatunnel-e2e/seatunnel-connector-v2-e2e/connector-kafka-e2e/src/test/java/org/apache/seatunnel/e2e/connector/kafka/KafkaIT.java]:
```suggestion
                .withLogConsumer(new Slf4jLogConsumer(DockerLoggerFactory.getLogger("confluentinc/cp-kafka:6.2.1")));
```

File: seatunnel-connectors-v2/connector-kafka/src/main/java/org/apache/seatunnel/connectors/seatunnel/kafka/sink/KafkaSinkWriter.java [hunk 12]
```diff
@@ -188,23 +207,32 @@ private void restoreState(List<KafkaSinkState> states) {
...
     }
+
+    private List<String> createPartitionKeys(Config pluginConfig, SeaTunnelRowType seaTunnelRowType) {
+        if (pluginConfig.hasPath(PARTITION_KEY_FIELDS)) {
+            return pluginConfig.getStringList(PARTITION_KEY_FIELDS).stream()
+                    .filter(f -> Arrays.asList(seaTunnelRowType.getFieldNames()).contains(f))
+                    .collect(Collectors.toList());
+        }
+        return null;
+    }
```
Comment (hailin0) [inline seatunnel-connectors-v2/connector-kafka/src/main/java/org/apache/seatunnel/connectors/seatunnel/kafka/sink/KafkaSinkWriter.java]: suggestion

```java
    private List<String> getPartitionKeyFields(Config pluginConfig, SeaTunnelRowType seaTunnelRowType) {
        if (pluginConfig.hasPath(PARTITION_KEY_FIELDS)) {
            List<String> partitionKeyFields = pluginConfig.getStringList(PARTITION_KEY_FIELDS);
            List<String> rowTypeFieldNames = Arrays.asList(seaTunnelRowType.getFieldNames());
            for (String partitionKeyField : partitionKeyFields) {
                if (!rowTypeFieldNames.contains(partitionKeyField)) {
                    throw new IllegalArgumentException(String.format(
                        "Partition key field not found: %s, rowType: %s", partitionKeyField, rowTypeFieldNames));
                }
            }
            return partitionKeyFields;
        }
        return Collections.emptyList();
    }
```

File: seatunnel-connectors-v2/connector-kafka/src/main/java/org/apache/seatunnel/connectors/seatunnel/kafka/sink/KafkaSinkWriter.java [hunk 13]
```diff
@@ -188,23 +207,32 @@ private void restoreState(List<KafkaSinkState> states) {
         }
     }
 
-    private Function<SeaTunnelRow, String> createPartitionExtractor(Config pluginConfig,
-                                                                    SeaTunnelRowType seaTunnelRowType) {
-        if (!pluginConfig.hasPath(PARTITION_KEY)){
+    private Function<SeaTunnelRow, SeaTunnelRow> createPartitionExtractor(SeaTunnelRowType seaTunnelRowType) {
```
Comment (hailin0) [inline seatunnel-connectors-v2/connector-kafka/src/main/java/org/apache/seatunnel/connectors/seatunnel/kafka/sink/KafkaSinkWriter.java]: suggestion

remove this method

File: seatunnel-connectors-v2/connector-kafka/src/main/java/org/apache/seatunnel/connectors/seatunnel/kafka/sink/KafkaSinkWriter.java [hunk 14]
```diff
@@ -162,12 +168,25 @@ private Properties getKafkaProperties(Config pluginConfig) {
 
     // todo: parse the target field from config
     private SeaTunnelRowSerializer<byte[], byte[]> getSerializer(Config pluginConfig, SeaTunnelRowType seaTunnelRowType) {
```
Comment (hailin0) [inline seatunnel-connectors-v2/connector-kafka/src/main/java/org/apache/seatunnel/connectors/seatunnel/kafka/sink/KafkaSinkWriter.java:146]: suggestion

```java
    private SeaTunnelRowSerializer<byte[], byte[]> getSerializer(Config pluginConfig, SeaTunnelRowType seaTunnelRowType) {
        if (pluginConfig.hasPath(PARTITION)){
            return new DefaultSeaTunnelRowSerializer(pluginConfig.getString(TOPIC),
                pluginConfig.getInt(PARTITION), seaTunnelRowType);
        } else {
            return new DefaultSeaTunnelRowSerializer(pluginConfig.getString(TOPIC),
                getPartitionKeyFields(pluginConfig, seaTunnelRowType), seaTunnelRowType);
        }
    }
```

File: seatunnel-connectors-v2/connector-kafka/src/main/java/org/apache/seatunnel/connectors/seatunnel/kafka/serialize/SeaTunnelRowSerializer.java [hunk 15]
```diff
@@ -34,9 +34,9 @@
     /**
      * Use Key serialize the {@link SeaTunnelRow} to a Kafka {@link ProducerRecord}.
      *
-     * @param key String
+     * @param key seatunnel row
      * @param row seatunnel row
      * @return kafka record.
      */
-    ProducerRecord<K, V> serializeRowByKey(String key, SeaTunnelRow row);
+    ProducerRecord<K, V> serializeRowByKey(SeaTunnelRow key, SeaTunnelRow row);
```
Comment (hailin0) [inline seatunnel-connectors-v2/connector-kafka/src/main/java/org/apache/seatunnel/connectors/seatunnel/kafka/serialize/SeaTunnelRowSerializer.java]: suggestion

remove this method
```suggestion
```

File: seatunnel-connectors-v2/connector-kafka/src/main/java/org/apache/seatunnel/connectors/seatunnel/kafka/sink/KafkaSinkWriter.java [hunk 16]
```diff
@@ -67,13 +72,12 @@ public class KafkaSinkWriter implements SinkWriter<SeaTunnelRow, KafkaCommitInfo
     // check config
     @Override
     public void write(SeaTunnelRow element) {
```
Comment (hailin0) [inline seatunnel-connectors-v2/connector-kafka/src/main/java/org/apache/seatunnel/connectors/seatunnel/kafka/sink/KafkaSinkWriter.java]: suggestion

```java
    public void write(SeaTunnelRow element) {
        ProducerRecord<byte[], byte[]> producerRecord = seaTunnelRowSerializer.serializeRow(element);
        kafkaProducerSender.send(producerRecord);
    }
```

File: seatunnel-connectors-v2/connector-kafka/src/main/java/org/apache/seatunnel/connectors/seatunnel/kafka/serialize/DefaultSeaTunnelRowSerializer.java [hunk 17]
```diff
@@ -25,34 +25,42 @@
 
 public class DefaultSeaTunnelRowSerializer implements SeaTunnelRowSerializer<byte[], byte[]> {
```
Comment (hailin0) [inline seatunnel-connectors-v2/connector-kafka/src/main/java/org/apache/seatunnel/connectors/seatunnel/kafka/serialize/DefaultSeaTunnelRowSerializer.java:31]: suggestion

```java
public class DefaultSeaTunnelRowSerializer implements SeaTunnelRowSerializer<byte[], byte[]> {

    private Integer partition;
    private final String topic;
    private final SerializationSchema keySerialization;
    private final SerializationSchema valueSerialization;

    public DefaultSeaTunnelRowSerializer(String topic, SeaTunnelRowType seaTunnelRowType) {
        this(topic, element -> null, createSerializationSchema(seaTunnelRowType));
    }

    public DefaultSeaTunnelRowSerializer(String topic, Integer partition, SeaTunnelRowType seaTunnelRowType) {
        this(topic, seaTunnelRowType);
        this.partition = partition;
    }

    public DefaultSeaTunnelRowSerializer(String topic,
                                         List<String> keyFieldNames,
                                         SeaTunnelRowType seaTunnelRowType) {
        this(topic, createKeySerializationSchema(keyFieldNames, seaTunnelRowType),
            createSerializationSchema(seaTunnelRowType));
    }

    public DefaultSeaTunnelRowSerializer(String topic,
                                         SerializationSchema keySerialization,
                                         SerializationSchema valueSerialization) {
        this.topic = topic;
        this.keySerialization = keySerialization;
        this.valueSerialization = valueSerialization;
    }

    @Override
    public ProducerRecord<byte[], byte[]> serializeRow(SeaTunnelRow row) {
        return new ProducerRecord<>(topic, partition,
            keySerialization.serialize(row), valueSerialization.serialize(row));
    }

    private static SerializationSchema createSerializationSchema(SeaTunnelRowType rowType) {
        return new JsonSerializationSchema(rowType);
    }

    private static SerializationSchema createKeySerializationSchema(List<String> keyFieldNames,
                                                                    SeaTunnelRowType seaTunnelRowType) {
        int[] keyFieldIndexArr = new int[keyFieldNames.size()];
        SeaTunnelDataType[] keyFieldDataTypeArr = new SeaTunnelDataType[keyFieldNames.size()];
        for (int i = 0; i < keyFieldNames.size(); i++) {
            String keyFieldName = keyFieldNames.get(i);
            int rowFieldIndex = seaTunnelRowType.indexOf(keyFieldName);
            keyFieldIndexArr[i] = rowFieldIndex;
            keyFieldDataTypeArr[i] = seaTunnelRowType.getFieldType(rowFieldIndex);
        }
        SeaTunnelRowType keyType = new SeaTunnelRowType(keyFieldNames.toArray(new String[0]), keyFieldDataTypeArr);
        SerializationSchema keySerializationSchema = createSerializationSchema(keyType);

        Function<SeaTunnelRow, SeaTunnelRow> keyDataExtractor = row -> {
            Object[] keyFields = new Object[keyFieldIndexArr.length];
            for (int i = 0; i < keyFieldIndexArr.length; i++) {
                keyFields[i] = row.getField(keyFieldIndexArr[i]);
            }
            return new SeaTunnelRow(keyFields);
        };
        return row -> keySerializationSchema.serialize(keyDataExtractor.apply(row));
    }
}
```

**Comment by harveyyue** [comment]
@hailin0 Great suggestions! Actually, you have refactor the sink serialization.

**Review by hailin0 — DISMISSED** [review]
LGTM

**Review by ic4y — DISMISSED** [review]
+1

**Comment by Hisoka-X** [comment]
Pleaes resolve conflict, Thanks!

**Comment by EricJoy2048** [comment]
Hi, @harveyyue . Resolve the conflicts please.

**Comment by harveyyue** [comment]
Sorry late response, have fixed conflicts.

**Comment by EricJoy2048** [comment]
Please merge the dev code after https://github.com/apache/incubator-seatunnel/pull/3401 merged.

**Review by TyrantLucifer — APPROVED** [review]
@hailin0 @Hisoka-X @EricJoy2048 PTAL.

**Review by TaoZex — APPROVED** [review]
+1

**Review by ic4y — APPROVED** [review]
LGTM

**Pull request merged**